### ISSUE TITLE
CLI: pass modules as argument

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#8740]
## Problem

I can't pass a module as argument to a function. This prevents a whole class of use cases that require one module dynamically loading another. For example, the "platform module":

- 100 different applications each have a Dagger module. They all implement the same interface: `Build()`, `Test()`
- 1 module defines an `App` interface which includes `Build` and `Test`; and it implements a reusable `Deploy` method which takes an array of the `App`s.
- App teams can call the platform module with *any version of their app module* in argument, and deploy it using a shared deployment pipeline. Crucially, no codegen or manual version pinning is required.

In effect, without this feature, a module cannot load another module without going through a static git-mediated update process *for each version of each loaded module*.

## Solution

Add a new feature to the `dagger` CLI: support for interface types as arguments.

If an argument is an interface type, interpret the command-line argument as a module address. Load the module; cast it as the argument's interface; then pass it. Voila!

**Comment by M-Pixel** [comment]
Is it possible to pass _any_ Dagger object other than the special cases like Directory as arguments in the CLI right now?  I've wanted to do

```
x = dagger core container from --address=something/distroless directory --path=/
dagger --interactive core container from --address=fedora with-directory --path=/inspect --directory=$x terminal
```

But couldn't find a way to do it mentioned in the documentation.  And I tried using from-id but the id expires by the time the next command runs.

Maybe the answer is, that's not how Dagger is supposed to be used, or that this should only be supported in dagger shell... and I think that would be perfectly reasonable.

**Comment by shykes** [comment]
@M-Pixel yeah the dagger shell is meant to make this kind of composition easier.

**Comment by shykes** [comment]
Your example in dagger shell:

```
x=$(container | from something/distroless)
container | from fedora | with-directory /inspect $x | terminal
```

**Comment by chrisjpalmer** [comment]
This kind of thing is basically what we (nine) are doing right now using dagger shell. Our setup is we are constructing "dependencies" of our "main" dagger module in dagger shell, then passing those dependencies into the main module. The main module accepts the dependencies by interface. 

An example of one of the dependencies we have is an AWS module that provides access to aws resources. We construct it in dagger shell to abstract away the differences per environment ( where our dagger functions are run ) during construction time. The CI environment relies on IMDS to initialise the module, where as local execution relies on the aws credentials file.

One thing that would really help to support this pattern is if dagger supported passing interface types to other interface types of other modules. To extend the above example, we have a post deploy tests module which needs to run some aws commands but it is invoked by the "main" module. Right now dagger does not support being able to do a `awsMod.AsPostDeployTestsAws()` from an interface type declared in the main module. We have a hacky workaround but it would be great if there was first class support for this :)
